outlines/fsm/json_schema.py

```diff
@@ -195,13 +195,7 @@ def to_regex(
             to_regex(resolver, t, whitespace_pattern) for t in instance["oneOf"]
         ]
 
-        xor_patterns = []
-        # json schema validation ensured there is no overlapping schemas in oneOf
-        for subregex in subregexes:
-            other_subregexes = filter(lambda r: r != subregex, subregexes)
-            other_subregexes_str = "|".join([f"{s}" for s in other_subregexes])
-            negative_lookahead = f"(?!.*({other_subregexes_str}))"
-            xor_patterns.append(f"({subregex}){negative_lookahead}")
+        xor_patterns = [f"(?:{subregex})" for subregex in subregexes]
```

Comment:
@lapp0 what is the difference bw oneOf and anyOf in this script? Both concat the subregex with `|`

Reply:
Do you also think additional conditions like just checking for `^` and just checking for `$` needs to be added to this conditon? https://github.com/outlines-dev/outlines/blob/7863f8e8bbaeb71c9d2434636a2d63bfe6dd7d39/outlines/fsm/json_schema.py#L248-L249

Reply:
And how about adding `array` along with `object` to the types to ignore [here](https://github.com/outlines-dev/outlines/blob/7863f8e8bbaeb71c9d2434636a2d63bfe6dd7d39/outlines/fsm/json_schema.py#L344-L347)? Array like object requires many more parameters like `items` which is not provided in the list of types which is the reason why object is ignored I believe since properties cannot be mentioned when type is given as a list

Reply:
@ekagra-ranjan actually reviewing the test cases, I think `anyOf` is wrong. I'd appreciate a second set of eyes on this. Would you agree that 

`oneOf` is correct (uses exclusive or), guarantees only one match

```
>>> print(re.fullmatch(r"((?:foo)|(?:bar))", "foo"))
<re.Match object; span=(0, 3), match='foo'>
>>> print(re.fullmatch(r"((?:foo)|(?:bar))", "foobar"))
None
```

`anyOf` is incorrect, allows only one match

```
>>> print(re.fullmatch(r"((foo)|(bar))", "foo"))
<re.Match object; span=(0, 3), match='foo'>
>>> print(re.fullmatch(r"((foo)|(bar))", "foobar"))
None
```

`allOf` is **somewhat** correct, requires generation of all schemas, but enforces an order which it shouldn't:

```
    elif "allOf" in instance:
        subregexes = [
            to_regex(resolver, t, whitespace_pattern) for t in instance["allOf"]
        ]
        subregexes_str = [f"{subregex}" for subregex in subregexes]
        return rf"({''.join(subregexes_str)})"
```

Reply:
For your other two comments, I appreciate your insights into json schema  -> regex. It's definitely a work in progress and it's great that you're looking deep into it's behavior and identifying problems. I think it's out of scope with this PR. Could you create a separate issue so others can provide input as well?